### PR TITLE
Option To Disable Record Data Health Check

### PR DESCRIPTION
**Pull request by andrewyu-salesforce**
Updated Custom Label:
- healthLabelDisableRecordBatchDescription

# Critical Changes

# Changes
 - When you run Health Check in NPSP Settings you now have the option to skip the record health checks and only run the configuration health checks.

# Issues Closed

# Community Ideas Delivered

# New Metadata

- Custom Field:
    - Error_Settings__c.DisableRecordDataHealthChecks__c

# Deleted Metadata

### CODE REVIEW

**Comment by andrewyu-salesforce** [comment]
Attach to [W-7782159](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008EIHvIAO/view)

**Comment by andrewyu-salesforce** [comment]
@sam-knox The label for Checkbox description is in line 7749 of customLabel file
Screen Shot 2020-08-20 at 10 02 02 AM

File: src/classes/STG_SettingsManager_TEST.cls [hunk 1]
```diff
@@ -1915,6 +1941,34 @@ private with sharing class STG_SettingsManager_TEST {
...
+        Test.startTest();
+        panel.runDetector();
+        Test.stopTest();
+
+        List<STG_PAnelHealthCheck_CTRL.DetectResult> results = panel.listDR;
+        Boolean hasDisabledHealthCheckWarningMessage = false;
+
+        for (STG_PAnelHealthCheck_CTRL.DetectResult result : results) {
+            if (result.strStatus == STG_PanelHealthCheck_CTRL.statusWarning && result.strName == System.Label.healthLabelDataBatchNotRun) {
+                hasDisabledHealthCheckWarningMessage = true;
```
Comment (dcaballeroSf) [inline src/classes/STG_SettingsManager_TEST.cls:1963]: this could be changed to hasDissabledHealthCheckWarningMessage = result.strStatus == STG(...)

**Comment by sam-knox** [comment]
@RadSFDeveloper - Release note Changes section updated.

File: src/classes/STG_SettingsManager_TEST.cls (hunk 1 above)
Comment (andrewyu-salesforce) [reply src/classes/STG_SettingsManager_TEST.cls:1963]: This will not work because there may be multiple results coming back and I do not want the boolean flag to be set to false because of other result

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 2]
```diff
@@ -412,15 +483,15 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
                     string.format(Label.healthDetailsBadReportId, new string[]{rptId}),
                     string.format(label.healthSolutionEditSetting, new string[]{strField, Label.stgNavPeople, Label.stgNavHouseholds}));
                 return;
             }
         } else {
-            list<Report> listRpt = [select id from Report where DeveloperName = 'NPSP_Campaign_Household_Mailing_List'];
+            List<Report> listRpt = [select id from Report where DeveloperName = 'NPSP_Campaign_Household_Mailing_List'];
             if (listRpt.size() == 0) {
                 createDR(strField, statusError, Label.healthDetailsMailingListReportMissing,
                    string.format(Label.healthSolutionMailingListReportMissing, new string[]{Label.stgNavPeople, Label.stgNavHouseholds}));
```
[the remark this answers is not in the document]
Comment (andrewyu-salesforce) [reply src/classes/STG_PanelHealthCheck_CTRL.cls]: I updated the entire class to follow the coding standard

**Comment by andrewyu-salesforce** [comment]
@diego-caballero Thanks for the review. The code is ready for another review

**Comment by dcaballeroSf** [comment]
@andrewyu-salesforce LGTM

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 3]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+    * @description Override saveSettings from STG_Panel to update Custom Setting Error Setting
+    * @return PageReference 
+    */
+    public override PageReference saveSettings() {
+        STG_SettingsManager_CTRL.idPanelCurrent = idPanel();
+        Savepoint sp = Database.setSavepoint();
+        try {
+            List<SObject> toUpdate = new List<SObject>();
+
+            if (STG_Panel.stgService.stgErr != null && !Test.isRunningTest()) {
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:75]: Should the check and update be done after `hasUpdatePermission` is validated?
```
            if (STG_Panel.stgService.stgErr != null && !Test.isRunningTest()) {
                update STG_Panel.stgService.stgErr;
            }
```

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 4]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+                toUpdate.add(STG_Panel.stgService.stgErr);
+            }
+
+            Boolean hasUpdatePermission = UTIL_Describe.getObjectDescribe(
+                String.valueOf(Error_Settings__c.SObjectType)
+            ).IsUpdateable();
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:81]: Methods start with a lower letter case: `isUpdateable();`

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 5]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+            if (!hasUpdatePermission) {
+                throw new DmlException(
+                    String.format(
+                        System.Label.flsError,
+                        new List<String>{System.Label.stgLabelErrorSettings})
+                );
+            }
+
+            update toUpdate;
+            listDR = new List<DetectResult>();
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:92]: Let's add a comment notifying that health check results will be cleared out when settings change.

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 6]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
         } private set;
     }
 
+    /*********************************************************************************************************
+    * @description Override saveSettings from STG_Panel to update Custom Setting Error Setting
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:66]: Let's provide more details what the Error Settings update impacts. It is not clear from reading this method how the Error settings is related to the Health Check configuration.

File: src/labels/CustomLabels.labels [hunk 7]
```diff
@@ -11659,6 +11659,14 @@ Note: The NPSP Settings tab is visible in the Nonprofit Success Pack application
         <shortDescription>stgLabelErrorNotify</shortDescription>
         <value>Error Notification Settings</value>
     </labels>
+    <labels>
+        <fullName>stgLabelErrorSettings</fullName>
+        <categories>Settings</categories>
+        <language>en_US</language>
+        <protected>false</protected>
```
Comment (RadSFDeveloper) [inline src/labels/CustomLabels.labels:11666]: If the new label should be created, it should be protected.

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 8]
```diff
@@ -180,28 +213,66 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
         STG_PanelOpps_CTRL.verifyOpportunityStageForPaymentAutoCloseIsActiveClosedWon(this);
         verifyScheduledJobs();
         STG_PanelERR_CTRL.verifyErrorSettings(this);
-        verifyOppContactRoles();
         verifyOppContactRolesSettings();
         verifyTriggerHandlers();
-        verifyNoMissingOppPayments();
         verifyLevels();
 
+        if (STG_Panel.stgService.stgErr.Disable_Record_Data_Health_Checks__c == false) {
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:220]: In the case the value is null, data record health check should not execute. Thus, let's check for `== true`.
Let's rename `generateDisableRecordHealthCheckMessages()` to `addDisableRecordHealthCheckWarnings()` or `addDisableRecordHealthCheckMessages()`.
```
        if (STG_Panel.stgService.stgErr.DisableRecordDataHealthChecks__c == true) {
            addDisableRecordHealthCheckWarnings();
        } else {
            executeDataRecordsHealthChecks();
        }
```

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 9]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+                        new List<String>{System.Label.stgLabelErrorSettings})
+                );
+            }
+
+            update toUpdate;
+            listDR = new List<DetectResult>();
+
+        } catch (Exception e) {
+            Database.rollback(sp);
+            ERR_Handler.processError(e, ERR_Handler_API.Context.STTG);
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:96]: Will `ERR_Handler.processError` add error message to the VF page? Check that only one error message is added to the VF page.

File: src/classes/STG_SettingsManager_TEST.cls [hunk 10]
```diff
@@ -1186,6 +1186,26 @@ private with sharing class STG_SettingsManager_TEST {
         Test.stopTest();
     }
 
+    /**
+    * @description Verifies that the user cannot update Error Settings without update permission 
+    */
+    @isTest
+    private static void shouldPreventUpateErrorSettingsWhenUserDoesNotHaveUpdatePermission() {
```
Comment (RadSFDeveloper) [inline src/classes/STG_SettingsManager_TEST.cls:1193]: Typo in the test name.

File: src/classes/STG_SettingsManager_TEST.cls [hunk 11]
```diff
@@ -1827,8 +1847,14 @@ private with sharing class STG_SettingsManager_TEST {
     @isTest
     private static void testPanelHealthCheck() {
         STG_PanelHealthCheck_CTRL panel = new STG_PanelHealthCheck_CTRL();
-        system.assertEquals('idPanelHealthCheck', panel.idPanel());
-        system.assertEquals(null, panel.runDetector());
+        
+        System.assertEquals('idPanelHealthCheck', panel.idPanel());
+        System.assertEquals(null, panel.runDetector());
+
+        STG_Panel.stgService.stgErr.Disable_Record_Data_Health_Checks__c = true;
```
Comment (RadSFDeveloper) [inline src/classes/STG_SettingsManager_TEST.cls:1854]: Should saving settings be another test, wondering why it was added to an existing test? Or perhaps the `panel.runDetector()` should be executed after settings are saved?

File: src/translations/de.translation [hunk 12]
```diff
@@ -3922,10 +3922,6 @@ Wenn Sie es vorziehen, können Sie &quot;Schenkungseingabe&quot; deaktivieren un
         <label>Statusprüfungen für Daten in Datensätzen deaktivieren</label>
         <name>healthLabelDisableRecordBatch</name>
     </customLabels>
-    <customLabels>
-        <label>Wenn sie aktiviert ist, prüft die Statusprüfung lediglich Ihre NPSP-Konfigurationseinstellungen auf Probleme.</label>
```
Comment (RadSFDeveloper) [inline src/translations/de.translation:3926]: @force2b , should we remove translations for the changed custom label, or update the label as following so it is updated by the translation team later on?
```
<label><!-- healthLabelDisableRecordBatchDescription --></label>
```

Comment (force2b) [reply src/translations/de.translation:3926]: Great question. Looking at the label change, it's not a fundamental change in the meaning of the text. It's really just adding an extra sentence to the end. Given that I think we should leave the current translation in place. It'll get automatically updated in the next translation round.

File: src/classes/STG_SettingsManager_TEST.cls (hunk 11 above)
Comment (andrewyu-salesforce) [reply src/classes/STG_SettingsManager_TEST.cls:1854]: When writing the test, I notice there are a series of tests calles testPanel....(), and for all of them, it's testing the main functionality of each NPSP page. That's why I added this chunk of code into this part. It is technically a duplicate of `shouldCreateWarningRecordsWhenDisabledRecordHealthCheckIsEnabled`. Will remove it

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 13]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+            if (STG_Panel.stgService.stgErr != null && !Test.isRunningTest()) {
+                toUpdate.add(STG_Panel.stgService.stgErr);
+            }
+
+            Boolean hasUpdatePermission = UTIL_Describe.getObjectDescribe(
+                String.valueOf(Error_Settings__c.SObjectType)
+            ).IsUpdateable();
+
+            if (!hasUpdatePermission) {
+                throw new DmlException(
```
[the remark this answers is not in the document]
Comment (andrewyu-salesforce) [reply src/classes/STG_PanelHealthCheck_CTRL.cls:84]: I was thinking of DmlException because the whole function is technically the validation part of the DML transaction. Will change to custom 'PermissionException' for this one.

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 14]
```diff
@@ -62,6 +62,44 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+                    String.format(
+                        System.Label.flsError,
+                        new List<String>{System.Label.stgLabelErrorSettings})
```
[the remark this answers is not in the document]
Comment (andrewyu-salesforce) [reply src/classes/STG_PanelHealthCheck_CTRL.cls:87]: I was thinking of Translation. If we use `getLabel()`. Then it will not be translated to other language because Custom Settings cannot be translate. Not sure which one is better. I also checked the Custom Label and it seem like we have almost Custom Labels for all other Custom Settings that were updated in NPSP Setting

File: src/classes/STG_PanelHealthCheck_CTRL.cls (hunk 9 above)
Comment (andrewyu-salesforce) [reply src/classes/STG_PanelHealthCheck_CTRL.cls:96]: `ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.ERROR, e.getMessage()));`(next line) adds the exception message to the error page. ERR_Handler.processError creates Error records in the error log.

**Comment by andrewyu-salesforce** [comment]
@RadSFDeveloper Thanks for the review. The code is updated and is ready for second review. I have couple responds/questions on certain feedback you gave.

File: src/classes/STG_PanelHealthCheck_CTRL.cls (hunk 9 above)
Comment (andrewyu-salesforce) [reply src/classes/STG_PanelHealthCheck_CTRL.cls:96]: Confirmed that only 1 message will be displayed onto the page and 1 error record will be created in error log.

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 15]
```diff
@@ -62,6 +62,45 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
         } private set;
     }
 
+    /*********************************************************************************************************
+    * @description Override saveSettings from STG_Panel to update Custom Setting Error Setting
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:66]: Let's replace description with 
```
* @description Save Error Settings "Disable Record Data Health Checks" checkbox related to the data Health Check verification.
```

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 16]
```diff
@@ -62,6 +62,45 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
+        } catch (Exception e) {
+            Database.rollback(sp);
+            ERR_Handler.processError(e, ERR_Handler_API.Context.STTG);
+            ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.ERROR, e.getMessage()));
+        }
+
+        return null;
+    }
+
+    public class NoPermissionException extends Exception {}
```
Comment (RadSFDeveloper) [inline src/classes/STG_PanelHealthCheck_CTRL.cls:103]: `NoPermissionException` -> `HealthCheckException` since we can generate other exceptions that are not related to permission checks.

File: src/objects/Error_Settings__c.object [hunk 17]
```diff
@@ -23,6 +23,16 @@
         <trackTrending>false</trackTrending>
         <type>Checkbox</type>
     </fields>
+    <fields>
+        <fullName>DisableRecordDataHealthChecks__c</fullName>
```
Comment (RadSFDeveloper) [inline src/objects/Error_Settings__c.object:27]: Since this new field is added, need to change [`UTIL_CustomSettingsFacade.getErrorSettingsForTests()`](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/UTIL_CustomSettingsFacade.cls#L844) to set the new field.
